File: libs/jgdsl/src/gdsl/asm/operand/Composite.java
package gdsl.asm.operand;

public class Composite extends Operand {
  private Operand[] operands;
  
  public Operand[] getOperands() {
    return operands;
  }
  
  public Composite (Operand[] operands) {
    this.operands = operands;
  }

  @Override public String toString () {
    StringBuilder sB = new StringBuilder();
    for (int i = 0; i < operands.length; i++) {
      if(i > 0)
        sB.append(":");
      sB.append(operands[i]);
    }
    return sB.toString();
  }

}

File: libs/cppgdsl/src/assembly/operand/composite.cpp
#include "cppgdsl/assembly/operand/composite.h"
#include "cppgdsl/assembly/operand/operand_visitor.h"
#include "cppgdsl/compare.h"

namespace gdsl {
namespace assembly {

void composite::put(std::ostream& out) const {
  out << "[";
  bool first = true;
  for (auto const& opnd : operands) {
    if (first)
      first = false;
    else
      out << ", ";
    out << *opnd;
  }
  out << "]";
}

void composite::accept(operand_visitor& ov) const { ov.visit(this); }

bool composite::operator==(operand const& o) const {
  bool is_equal = false;
  operand_visitor v(true);
  v._([&](composite const* o) {
    is_equal = equals(this->operands, o->operands);
  });
  o.accept(v);
  return is_equal;
}

}  // namespace assembly
}  // namespace gdsl
